In [ ]:
# pytutorial-1

##Chapter 1  The way of the program

The goal of this book is to teach you to think like a computer scientist. This way of thinking combines some of the best features of mathematics, engineering, and natural science. Like scientists, they observe the behavior of complex systems, form hypotheses, and test predictions.

The single most important skill for a computer scientist is _problem solving_. Problem solving means the ability to formulate problems, think creatively about solutions, and express a solution clearly and accurately. As it turns out, the process of learning to program is an excellent opportunity to practice problem-solving skills. That’s why this chapter is called, “The way of the program”.

On one level, you will be learning to program, a useful skill by itself. On another level, you will use programming as a means to an end. As we go along, that end will become clearer.

**1.1  What is a program?**

A **program** is a sequence of instructions that specifies how to perform a computation. 
The computation might be:
* something mathematical(solving a system of equations or finding the roots of a polynomial)
* a symbolic computation (searching and replacing text in a document or something graphical, like processing an image or playing a video)

The details look different in different languages, but a few basic instructions appear in just about every language:

**input:**
Get data from the keyboard, a file, the network, or some other device.
**output:**
Display data on the screen, save it in a file, send it over the network, etc.
**math:**
Perform basic mathematical operations like addition and multiplication.
**conditional execution:**
Check for certain conditions and run the appropriate code.
**repetition:**
Perform some action repeatedly, usually with some variation.
__Believe it or not, that’s pretty much all there is to it.__ 
__So you can think of programming as the process of breaking a large, complex task into smaller and smaller subtasks until the subtasks are simple enough to be performed with one of these basic instructions.__


### 1.2  Running Python

* There are a number of web pages you can use to run Python. If you already have a favorite, go ahead and use it. Otherwise I recommend PythonAnywhere. I provide detailed instructions for getting started at http://tinyurl.com/thinkpython2e.

* There are two versions of Python, called Python 2 and Python 3. They are very similar, so if you learn one, it is easy to switch to the other. In fact, there are only a few differences you will encounter as a beginner. This book is written for Python 3, but I include some notes about Python 2.

* The **Python interpreter** is a program that _reads and executes_ Python code. Depending on your environment, you might start the interpreter by clicking on an icon, or by typing **python** on a command line. When it starts, you should see output like this:

```
> python
Python 3.4.0 (default, Jun 19 2015, 14:20:21) 
[GCC 4.8.2] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```
The first three lines contain information about the interpreter and the operating system it’s running on, so it might be different for you. But you should check that the version number, which is 3.4.0 in this example, begins with 3, which indicates that you are running Python 3. If it begins with 2, you are running (you guessed it) Python 2.

The last line **```** is a prompt that indicates that the interpreter is ready for you to enter code. If you type a line of code and hit Enter, the interpreter displays the result:

>>> 1 + 1
2

But, __python__ command didn't give a version, so I did.
```
> python -V
Python 2.7.14 :: Anaconda, Inc.
```


Now you’re ready to get started. From here on, I assume that you know how to start the Python interpreter and run code.


###1.3  The first program

Traditionally, the first program you write in a new language is called “Hello, World!” because all it does is display the words “Hello, World!”. In Python, it looks like this:

```
>>> print('Hello, World!')
Hello, World!
```
* The quotation marks in the program mark the beginning and end of the text to be displayed; they don’t appear in the result.
* The parentheses indicate that print is a function. We’ll get to functions in Chapter 3.

In Python 2, the print statement is slightly different; it is not a function, so it doesn’t use parentheses.

>>> print 'Hello, World!'
This distinction will make more sense soon, but that’s enough to get started.



### 1.4  Arithmetic operators

After “Hello, World”, the next step is **arithmetic**. Python provides **operators**, which are special symbols that represent computations like addition and multiplication.

The operators +, -, and * perform addition, subtraction, and multiplication, as in the following examples:



In [4]:
40+2

42

In [5]:
43-1

42

In [6]:
6*7

42

In [ ]:
The operator / performs division:

In [7]:
84/2


42

In [ ]:
* You might wonder why the result is 42.0 instead of 42. I’ll explain in the next section.

Finally, the operator ** performs **exponentiation**; that is, it raises a number to a power:


In [9]:
6**2 + 6


42

In [ ]:
* In some other languages, ^ is used for exponentiation, but in Python it is a bitwise operator called XOR. If you are not familiar with bitwise operators, the result will surprise you:


In [10]:
6^2

4

In [ ]:
I won’t cover bitwise operators in this book, but you can read about them at http://wiki.python.org/moin/BitwiseOperators.


### 1.5  Values and types

A **value** is one of the basic things a program works with, like a **letter or a number**. Some values we have seen so far are 2, 42.0, and 'Hello, World!'.

These values belong to different types: 2 is an **integer**, 42.0 is a **floating-point number**, and 'Hello, World!' is a **string**, so-called because **the letters it contains are strung together**.

If you are not sure what type a value has, the interpreter can tell you:

```
>>> type(2)
<class 'int'>
>>> type(42.0)
<class 'float'>
>>> type('Hello, World!')
<class 'str'>
```

* In these results, the word “class” is used in the sense of a **category**; a type is a **category of values**.

    * integers = type int
    * strings = str
    * floating-point numbers = float

* What about values like '2' and '42.0'? They look like numbers, but they are in **quotation marks** like **strings**.

```
>>> type('2')
<class 'str'>
>>> type('42.0')
<class 'str'>
They’re strings.

```

* Use commas between groups of digits, as in 1,000,000 is **not** a legal integer in Python, but it is legal:

```
>>> 1,000,000
(1, 0, 0)
That’s not what we expected at all! Python interprets 1,000,000 as a **comma-separated sequence of integers**. We’ll learn more about this kind of sequence later.
```

### 1.6  Formal and natural languages

* Natural languages are the languages people speak, such as English, Spanish, and Japanese. They were not designed by people (although people try to impose some order on them); **they evolved naturally**.

* **Formal languages** are languages that are designed by people for specific applications. For example, the notation that mathematicians use is a formal language that is particularly good at denoting relationships among numbers and symbols. Chemists use a formal language to represent the chemical structure of molecules. And most importantly:

* Programming languages are formal languages that have been designed to **express computations**. Formal languages tend to have strict syntax rules that govern the structure of statements. For example, in mathematics the statement 3 + 3 = 6 has correct syntax, but 3 + = 3 $ 6 does not. In chemistry H2O is a syntactically correct formula, but 2Zz is not.

* Syntax rules come in two flavors, pertaining to **tokens** and **structure**. **Tokens are the basic elements of the language**(words, numbers, and chemical elements). One of the problems with 3 += 3 $ 6 is that $ is not a legal token in mathematics (at least as far as I know). Similarly, 2Zz is not legal because there is no element with the abbreviation Zz.

* The second type of syntax rule pertains to **the way tokens are combined**. The equation 3 += 3 is illegal because even though + and = are legal tokens, you can’t have one right after the other. Similarly, in a chemical formula the subscript comes after the element name, not before.

* This is @ well-structured Engli$h sentence with invalid t*kens in it. This sentence all valid tokens has, but invalid structure with.

* When you read a sentence in English or a statement in a formal language, you have to figure out the **structure** (although in a natural language you do this subconsciously). This process is called *parsing**.

* Although formal and natural languages have many features in common—tokens, structure, and syntax—there are some differences:

**ambiguity:**
Natural languages are full of **ambiguity**, which people deal with by using contextual clues and other information. Formal languages are designed to be nearly or completely unambiguous, which means that any statement has exactly one meaning, regardless of context.

**redundancy:**
In order to make up for ambiguity and reduce misunderstandings, **natural languages employ lots of redundancy**. As a result, they are often verbose. **Formal languages are less redundant and more concise**.

**literalness:**
Natural languages are full of **idiom and metaphor**. If I say, “The penny dropped”, there is probably no penny and nothing dropping (this idiom means that someone understood something after a period of confusion). **Formal languages mean exactly what they say**.
Because we all grow up speaking natural languages, it is sometimes hard to adjust to formal languages. The difference between formal and natural language is like the difference between poetry and prose, but more so:

**Poetry:**
Words are used for their sounds as well as for their meaning, and the whole poem together creates an effect or emotional response. Ambiguity is not only common but often deliberate.

**Prose:**
The literal meaning of words is more important, and the structure contributes more meaning. Prose is more amenable to analysis than poetry but still often ambiguous.

**Programs:**
The meaning of a **computer program** is **unambiguous and literal**, and can be understood entirely by **analysis of the tokens and structure**.

* Formal languages are more **dense** than natural languages, so it takes longer to read them. Also, the structure is important, so it is not always best to read from top to bottom, left to right. Instead, learn to **parse** the program in your head, **identifying the tokens and interpreting the structure**. 
* Finally, the details matter. Small errors in spelling and punctuation, which you can get away with in natural languages, can make a big difference in a formal language.

### 1.7  Debugging

**Programmers make mistakes**. For whimsical reasons, **programming errors** are called **bugs** and the **process of tracking them down** is called **debugging**.

Programming, and especially debugging, sometimes brings out strong emotions. If you are struggling with a difficult bug, you might feel angry, despondent, or embarrassed.

There is evidence that people naturally respond to computers as if they were people. When they work well, we think of them as teammates, and when they are obstinate or rude, we respond to them the same way we respond to rude, obstinate people (Reeves and Nass, The Media Equation: How People Treat Computers, Television, and New Media Like Real People and Places).

Preparing for these reactions might help you deal with them. 
**One approach is to think of the computer as an employee with certain strengths**, like speed and precision, and particular weaknesses, like lack of empathy and inability to grasp the big picture.

Your job is to be a good manager: find ways to take advantage of the strengths and mitigate the weaknesses. And find ways to use your emotions to engage with the problem, without letting your reactions interfere with your ability to work effectively.

Learning to debug can be frustrating, but it is a valuable skill that is useful for many activities beyond programming. At the end of each chapter there is a section, like this one, with my suggestions for debugging. I hope they help!


### 1.8  Glossary

**problem solving:**
The process of formulating a problem, finding a solution, and expressing it.

**high-level language:**
A programming language like Python that is designed to be easy for **humans to read and write**.

**low-level language:**
A programming language that is designed to be **easy for a computer to run**; also called “machine language” or “assembly language”.

**portability:**
A property of a program that can run on more than one kind of computer.

**interpreter:**
A program that reads another program and executes it

**prompt:**
Characters displayed by the interpreter to indicate that it is ready to take input from the user.

**program:**
A set of instructions that specifies a computation.

**print statement:**
An instruction that causes the Python interpreter to display a value on the screen.

**operator:**
A special symbol that represents a simple computation like addition, multiplication, or string concatenation.

**value:**
One of the basic units of data, like a number or string, that a program manipulates.

**type:**
A category of values. The types we have seen so far are integers (type int), floating-point numbers (type float), and strings (type str).

**integer:**
A type that represents whole numbers.

**floating-point:**
A type that represents numbers with fractional parts.

**string:**
A type that represents sequences of characters.

**natural language:**
Any one of the languages that people speak that evolved naturally.

**formal language:**
Any one of the languages that people have designed for specific purposes, such as representing mathematical ideas or computer programs; all programming languages are formal languages.

**token:**
One of the basic elements of the syntactic structure of a program, analogous to a word in a natural language.

**syntax:**
The rules that govern the structure of a program.

**parse:**
To examine a program and analyze the syntactic structure.

**bug:**
An error in a program.

**debugging:**
The process of finding and correcting bugs.


### 1.9  Exercises

**Exercise 1**  
It is a good idea to read this book in front of a computer so you can try out the examples as you go.

Whenever you are experimenting with a new feature, you should try to make mistakes. 

* In the “Hello, world!” program, what happens if you leave out one of the quotation marks? 
```>>> print 'Hello World!
  File "<stdin>", line 1
    print 'Hello World!
                      ^
SyntaxError: EOL while scanning string literal
>>> 
```
* What if you leave out both? What if you spell print wrong?

```
>>> print Hello World!
  File "<stdin>", line 1
    print Hello World!
                    ^
SyntaxError: invalid syntax
>>> 

```
This kind of experiment helps you remember what you read; it also helps when you are programming, because you get to know what the error messages mean. It is better to make mistakes now and on purpose than later and accidentally.

* In a print statement, what happens if you leave out one of the parentheses, or both?
If you are trying to print a string, what happens if you leave out one of the quotation marks, or both?
```
>>> print ('Hello World!')
Hello World!
>>> print ('Hello World!'
... 
KeyboardInterrupt
>>> print 'Hello World!'
Hello World!
>>> 
```

* You can use a minus sign to make a negative number like -2. 
```
>>> 2+-2
0
```

* What happens if you put a plus sign before a number? What about 2++2?
```
>>> +2+2
4
>>> 2++2
4
```

* In math notation, leading zeros are ok, as in 02. What happens if you try this in Python?
```
>>> 02+02
4
```

* What happens if you have two values with no operator between them?
```
>>> 22
22
```



**Exercise 2**

Start the Python interpreter and use it as a calculator.

* How many seconds are there in 42 minutes 42 seconds?


In [11]:
42*60+42

2562

* How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.
x/10km=1mi/1.61km

In [12]:
a=10/1.61
print(round(a, 2))

6.21


In [ ]:
* If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace (time per mile in minutes and seconds)? 

Time: 42 minutes 42 seconds = 2562 seconds
Distance: 10 km = 6.2 miles

Pace= Time/Distance (mile)= 2562 seconds/6.2 miles = 413.226 sec/mile = 413.226/60= 6.9

In [13]:
a = 2562/6.2/60
print(round(a, 2))

6.89


__Answer: Pace:6 minutes and 9 seconds__

* What is your average speed in miles per hour?

Convert seconds to hours
2562.0 seconds/3600.00= 0.71 hours


In [14]:
a = 2562.0/3600.0
print(round(a, 2))

0.71


In [15]:
a = 6.2/0.71
print(round(a, 2))


8.73


Answer: Speed = Distance/Time = 6.2 miles/0.71 hours = 8.73 miles per hour

In [ ]:


# Chapter 2  Variables, expressions and statements

One of the most powerful features of a programming language is the ability to manipulate variables. A **variable** is a name that refers to a **value**.


### 2.1  Assignment statements

An assignment statement creates a new variable and gives it a value:

```
>>> message = 'And now for something completely different'
>>> n = 17
>>> pi = 3.141592653589793
```

This example makes three assignments. The first assigns a string to a new variable named **message**; the second gives the integer 17 to **n**; the third assigns the (approximate) value of π to **pi**.

A common way to represent variables on paper is to write the name with an arrow pointing to its value. This kind of figure is called a **state diagram** because it shows what state each of the variables is in (think of it as the variable’s state of mind). 

### 2.2  Variable names

Programmers generally choose names for their variables that are meaningful—they document what the variable is used for.

* Variable names can be as **long** as you like. 
* They can contain both **letters and numbers**, but **they can’t begin with a number**. 
* It is legal to use uppercase letters, but it is **conventional to use only lower case** for variables names.
* The underscore character, _, can appear in a name. It is often used in names with multiple words, such as your_name or airspeed_of_unladen_swallow.

If you give a variable an illegal name, you get a syntax error:

```
>>> 76trombones = 'big parade'
SyntaxError: invalid syntax
__76trombones is illegal because it begins with a number__

>>> more@ = 1000000
SyntaxError: invalid syntax
__more@ is illegal because it contains an illegal character, @.__

>>> class = 'Advanced Theoretical Zymurgy'
SyntaxError: invalid syntax
__It turns out that class is one of Python’s keywords. The interpreter uses keywords to recognize the structure of the program, and they cannot be used as variable names.__
```

Python 3 has these keywords:

False      class      finally    is         return
None       continue   for        lambda     try
True       def        from       nonlocal   while
and        del        global     not        with
as         elif       if         or         yield
assert     else       import     pass
break      except     in         raise

You don’t have to memorize this list. In most development environments, keywords are displayed in a different color; if you try to use one as a variable name, you’ll know.


### 2.3  Expressions and statements

An **expression** is a combination of values, variables, and operators. A value all by itself is considered an expression, and so is a variable, so the following are all legal expressions:

```
>>> 42
42
>>> n
17
>>> n + 25
42
```

When you type an expression at the prompt, the interpreter evaluates it, which means that it finds **the value of the expression**. In this example, n has the value 17 and n + 25 has the value 42.

A statement is **a unit of code that has an effect**, like creating a variable or displaying a value.

```
>>> n = 17
>>> print(n)
```

The first line is an **assignment statement** that gives a value to n. The second line is a **print statement** that displays the value of n.

When you type a **statement**, the **interpreter** executes it, which means that it does whatever the statement says. In general, statements don’t have values.


### 2.4  Script mode

So far we have run Python in interactive mode, which means that you interact directly with the interpreter. Interactive mode is a good way to get started, but if you are working with more than a few lines of code, it can be clumsy.

The alternative is to save code in a file called a **script** and then run the interpreter in script mode to execute the script. By convention, Python scripts have names that end with **.py.**

If you know how to create and run a script on your computer, you are ready to go. Otherwise I recommend using PythonAnywhere again. I have posted instructions for running in script mode at http://tinyurl.com/thinkpython2e.

Because Python provides both modes, you can test bits of code in interactive mode before you put them in a script. But there are differences between interactive mode and script mode that can be confusing.

For example, if you are using Python as a calculator, you might type

```
>>> miles = 26.2
>>> miles * 1.61
42.182
```

The first line **assigns** a value to miles, but it has no visible effect. 
The second line is an **expression**, so the interpreter evaluates it and displays the result. It turns out that a marathon is about 42 kilometers.

But if you type the same code into a script and run it, you get no output at all. In script mode an expression, all by itself, has no visible effect. Python actually evaluates the expression, **but it doesn’t display the value unless you tell it to:**

```
miles = 26.2
print(miles * 1.61)
```
This behavior can be confusing at first.

A script usually contains a sequence of statements. If there is more than one statement, the results appear one at a time as the statements execute.

For example, the script

```
print(1)
x = 2
print(x)
produces the output

1
2
```
The assignment statement produces no output.

To check your understanding, type the following statements in the Python interpreter and see what they do:

```
5
x = 5
x + 1
```
Now put the same statements in a script and run it. What is the output? Modify the script by transforming each expression into a print statement and then run it again.

### 2.5  Order of operations

When an expression contains more than one operator, the order of evaluation depends on the order of operations. For mathematical operators, Python follows mathematical convention. The **acronym PEMDAS** is a useful way to remember the rules:

* **Parentheses** have the highest precedence and can be used to force an expression to evaluate in the order you want. Since expressions in parentheses are evaluated first, 2 * (3-1) is 4, and (1+1)**(5-2) is 8. You can also use parentheses to make an expression easier to read, as in (minute * 100) / 60, even if it doesn’t change the result.
* **Exponentiation** has the next highest precedence, so 1 + 2**3 is 9, not 27, and 2 * 3**2 is 18, not 36.
* **Multiplication** and Division have higher precedence than Addition and Subtraction. So 2*3-1 is 5, not 4, and 6+4/2 is 8, not 5.
* **Operators** with the same precedence are evaluated from left to right (except exponentiation). So in the expression degrees / 2 * pi, the division happens first and the result is multiplied by pi. To divide by 2 π, you can use parentheses or write degrees / 2 / pi.


### 2.6  String operations

In general, you can’t perform mathematical operations on strings, even if the strings look like numbers, so the following are illegal:

'2'-'1'    'eggs'/'easy'    'third'*'a charm'
**But there are two exceptions, + and *.  **

The + operator performs string concatenation, which means it joins the strings by linking them end-to-end. For example:

```
>>> first = 'throat'
>>> second = 'warbler'
>>> first + second
throatwarbler
```

**The * operator also works on strings**; it performs repetition. 
For example, 'Spam'*3 is 'SpamSpamSpam'. 
If one of the values is a string, the other has to be an integer.

This use of + and * makes sense by analogy with addition and multiplication. Just as 4*3 is equivalent to 4+4+4, we expect 'Spam'*3 to be the same as 'Spam'+'Spam'+'Spam', and it is. 


### 2.7  Comments

As programs get bigger and more complicated, they get more difficult to read. Formal languages are dense, and it is often difficult to look at a piece of code and figure out what it is doing, or why.

For this reason, it is a good idea to add **notes** to your programs to explain in natural language what the program is doing. 

These notes are called comments, and they start with the # symbol:

# compute the percentage of the hour that has elapsed
``
percentage = (minute * 100) / 60
``
In this case, the comment appears on a line by itself. You can also put comments at the end of a line:
```
percentage = (minute * 100) / 60     # percentage of an hour
```
Everything from the # to the end of the line is ignored—it has no effect on the execution of the program.


This comment is redundant with the code and useless:

```
v = 5     # assign 5 to v
This comment contains useful information that is not in the code:
```
```
v = 5     # velocity in meters/second. 
Good variable names can reduce the need for comments, but long names can make complex expressions hard to read, so there is a tradeoff.
```

### 2.8  Debugging

Three kinds of errors can occur in a program: 
1. syntax errors
2. runtime errors
3. semantic errors

It is useful to distinguish between them in order to track them down more quickly.

**Syntax error:**
* “Syntax” refers to the **structure** of a program and the rules about that structure. For example, parentheses have to come in matching pairs, so (1 + 2) is legal, but 8) is a syntax error.
* If there is a syntax error anywhere in your program, Python displays an error message and quits, and you will not be able to run the program. 
* During the first few weeks of your programming career, you might spend a lot of time tracking down syntax errors. As you gain experience, you will make fewer errors and find them faster.

**Runtime error:**
* The second type of error is a runtime error, so called because the error does not appear until after the program has started running. 
* These errors are also called exceptions because they usually indicate that something exceptional (and bad) has happened.
* Runtime errors are rare in the simple programs you will see in the first few chapters, so it might be a while before you encounter one.

**Semantic error:**
* The third type of error is “semantic”, which means related to "meaning". 
* If there is a semantic error in your program, it will run without generating error messages, but it will not do the right thing. It will do something else. Specifically, it will do what you told it to do.
** Identifying semantic errors can be tricky because it requires you to work backward by looking at the output of the program and trying to figure out what it is doing.

### 2.9  Glossary

**variable:**
A name that refers to a value.

**assignment:**
A statement that assigns a value to a variable.

**state diagram:**
A graphical representation of a set of variables and the values they refer to.

**keyword:**
A reserved word that is used to parse a program; you cannot use keywords like if, def, and while as variable names.

**operand:**
One of the values on which an operator operates.

**expression:**
A combination of variables, operators, and values that represents a single result.

**evaluate:**
To simplify an expression by performing the operations in order to yield a single value.

**statement:**
A section of code that represents a command or action. So far, the statements we have seen are assignments and print statements.

**execute:**
To run a statement and do what it says.

**interactive mode:**
A way of using the Python interpreter by typing code at the prompt.

**script mode:**
A way of using the Python interpreter to read code from a script and run it.

**script:**
A program stored in a file.

**order of operations:**
Rules governing the order in which expressions involving multiple operators and operands are evaluated.

**concatenate:**
To join two operands end-to-end.

**comment:**
Information in a program that is meant for other programmers (or anyone reading the source code) and has no effect on the execution of the program.

**syntax error:**
An error in a program that makes it impossible to parse (and therefore impossible to interpret).

**exception:**
An error that is detected while the program is running.

**semantics:**
The meaning of a program.

**semantic error:**
An error in a program that makes it do something other than what the programmer intended.


### 2.10  Exercises

#### Exercise 1  

Repeating my advice from the previous chapter, whenever you learn a new feature, you should try it out in interactive mode and make errors on purpose to see what goes wrong.

We’ve seen that n = 42 is legal. What about 42 = n?
How about x = y = 1?

In [16]:
42 = n


SyntaxError: can't assign to literal (<ipython-input-16-600427bfcdb2>, line 1)

In [17]:
x = y = 1 

In [18]:
x

1

In [19]:
y

1

In [ ]:
In some languages every statement ends with a semi-colon, ;. What happens if you put a semi-colon at the end of a Python statement?

What if you put a period at the end of a statement?

In math notation you can multiply x and y like this: x y. What happens if you try that in Python?


In [22]:
x = 2;

x

In [26]:
x = 3.

In [27]:
x

3.0

In [28]:
xy

NameError: name 'xy' is not defined


#### Exercise 2  

Practice using the Python interpreter as a calculator:

1. The volume of a sphere with radius r is 4/3 π r3. What is the volume of a sphere with radius 5?


In [33]:
p = 3.14
(4/3)*p*(5**3)

392.5

2. Suppose the cover price of a book is $24.95, but bookstores get a 40% discount. Shipping costs $3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [39]:
x = ((24.95*60) - (0.75*59) - 3)/0.60
print(round(x, 2))


2416.25


3. If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast?


In [42]:
(8*2+21)+(15*2+36)/60

38

It takes 38 min for the person to come home.

In [43]:
38+52


90

6 o'clock + 90 min 
Answer: The person gets home at 7:30am.

In [ ]:


## Chapter 3  Functions

In the context of programming, a function is a named sequence of statements that performs a computation. When you define a function, you specify the name and the sequence of statements. Later, you can “call” the function by name.

#### 3.1  Function calls

We have already seen one example of a function call:

```
>>> type(42)
<class 'int'>
The name of the function is type. The expression in parentheses is called the argument of the function. The result, for this function, is the type of the argument.
```

It is common to say that a function “takes” an argument and “returns” a result. The result is also called the return value.

Python provides functions that convert values from one type to another. The int function takes any value and converts it to an integer, if it can, or complains otherwise:

```
>>> int('32')
32
>>> int('Hello')
ValueError: invalid literal for int(): Hello
int can convert floating-point values to integers, but it doesn’t round off; it chops off the fraction part:

>>> int(3.99999)
3
>>> int(-2.3)
-2
float converts integers and strings to floating-point numbers:

>>> float(32)
32.0

>>> float('3.14159')
3.14159
Finally, str converts its argument to a string:

>>> str(32)
'32'
>>> str(3.14159)
'3.14159'
```

#### 3.2  Math functions


Python has a math module that **provides most of the familiar mathematical functions**. A module is a file that contains a collection of related functions.

Before we can use the functions in a module, **we have to import it with an import statement**:


**>>> import math**
This statement creates a module object named math. If you display the module object, you get some information about it:

>>> math
<module 'math' (built-in)>
e.g. <module 'math' from '/Users/teachinghouse/miniconda/lib/python2.7/lib-dynload/math.so'>

The module object contains the functions and variables defined in the module. To access one of the functions, **you have to specify the name of the module and the name of the function, separated by a dot (also known as a period). This format is called dot notation**.

e.g. math.log10  nameofmodule.nameoffunction
>>> ratio = signal_power / noise_power
>>> decibels = 10 * math.log10(ratio)

The first example uses math.log10 to compute a signal-to-noise ratio in decibels (assuming that signal_power and noise_power are defined). The math module also provides log, which computes logarithms base e.

>>> radians = 0.7
>>> height = math.sin(radians)

The second example finds the sine of radians. The name of the variable is a hint that sin and the other trigonometric functions (cos, tan, etc.) take arguments in radians. To convert from degrees to radians, divide by 180 and multiply by π:

>>> degrees = 45
>>> radians = degrees / 180.0 * math.pi
>>> math.sin(radians)
0.707106781187

The expression **math.pi** gets the variable pi from the math module. Its value is a floating-point approximation of π, accurate to about 15 digits.

If you know trigonometry, you can check the previous result by comparing it to the square root of two divided by two:

>>> math.sqrt(2) / 2.0
0.707106781187


#### 3.3  Composition

So far, we have looked at the elements of a program—variables, expressions, and statements—in isolation, without talking about **how to combine them**.

One of the most useful features of programming languages is their ability to take small building blocks and compose them. 

For example, the argument of a function can be any kind of expression, including arithmetic operators:

x = math.sin(degrees / 360.0 * 2 * math.pi)

And even function calls:

x = math.exp(math.log(x+1))

Almost anywhere you can put a value, you can put an arbitrary expression, with one exception: the left side of an assignment statement has to be a variable name. Any other expression on the left side is a syntax error (we will see exceptions to this rule later).

>>> minutes = hours * 60                 # right
>>> hours * 60 = minutes                 # wrong!
SyntaxError: can't assign to operator


### 3.4  Adding new functions

So far, we have only been using the functions that come with Python, but it is also possible to add new functions. A **function** definition specifies the name of a new function and the sequence of statements that run when the function is called.

Here is an example:

def print_lyrics():
    print("I'm a lumberjack, and I'm okay.")  #definning the header
    print("I sleep all night and I work all day.") #definninf the body
    
* **def** is a keyword that indicates that this is a function definition. The **name of the function** is print_lyrics. 
* The rules for function names are the same as for variable names: letters, numbers and underscore are legal, but the first character can’t be a number. You can’t use a keyword as the name of a function, and you should avoid having a variable and a function with the same name.
* The **empty parentheses** after the name indicate that this function doesn’t take any arguments.
* The **first line** of the function definition is called **the header**; **the rest** is called **the body**. The header has to end with a colon and the body has to be indented. By convention, indentation is always four spaces. The body can contain any number of statements.
* The strings in the print statements are enclosed in **double quotes**. **Single quotes and double quotes do the same thing**; most people use single quotes except in cases like this where a single quote (which is also an apostrophe) appears in the string.
* All quotation marks (single and double) must be “straight quotes”, usually located next to Enter on the keyboard. “Curly quotes”, like the ones in this sentence, are not legal in Python.

If you type a function definition in interactive mode, the interpreter prints dots (...) to let you know that the definition isn’t complete:

>>> def print_lyrics():
...     print("I'm a lumberjack, and I'm okay.")
...     print("I sleep all night and I work all day.")
...
To end the function, **you have to enter an empty line**.

Defining a function creates a function object, which has type function:

>>> print(print_lyrics)
<function print_lyrics at 0xb7e99e9c>
>>> type(print_lyrics)
<class 'function'>
The syntax for calling the new function is the same as for built-in functions:

>>> print_lyrics()
I'm a lumberjack, and I'm okay.
I sleep all night and I work all day.

Once you have defined a function, you can use it **inside another function**. For example, to repeat the previous refrain, we could write a function called repeat_lyrics:

def repeat_lyrics():
    print_lyrics()
    print_lyrics()
And then call repeat_lyrics:

>>> repeat_lyrics()
I'm a lumberjack, and I'm okay.
I sleep all night and I work all day.
I'm a lumberjack, and I'm okay.
I sleep all night and I work all day.
But that’s not really how the song goes.


### 3.5  Definitions and uses

Pulling together the code fragments from the previous section, the whole program looks like this:

def print_lyrics():
    print("I'm a lumberjack, and I'm okay.")
    print("I sleep all night and I work all day.")

def repeat_lyrics():
    print_lyrics()
    print_lyrics()

repeat_lyrics()

* This program contains two function definitions: print_lyrics and repeat_lyrics. Function definitions get executed just like other statements, but the effect is to create function objects. The statements inside the function do not run until the function is called, and the function definition generates no output.

* As you might expect, you have to create a function before you can run it. In other words, the function definition has to run before the function gets called.
* As an exercise, move the last line of this program to the top, so the function call appears before the definitions. Run the program and see what error message you get.

In [44]:

repeat_lyrics()

def print_lyrics():
    print("I'm a lumberjack, and I'm okay.")
    print("I sleep all night and I work all day.")

def repeat_lyrics():
    print_lyrics()
    print_lyrics()



NameError: name 'repeat_lyrics' is not defined

* Now move the **function call** back to the bottom and move the definition of print_lyrics after the definition of repeat_lyrics. What happens when you run this program?

In [45]:
def print_lyrics():
   
def repeat_lyrics():
    print("I'm a lumberjack, and I'm okay.")
    print("I sleep all night and I work all day.")
    print_lyrics()
    print_lyrics()

repeat_lyrics()

IndentationError: expected an indented block (<ipython-input-45-0fa99cdff977>, line 3)

### 3.6  Flow of execution

To ensure that a function is defined before its first use, you have to know the order statements run in, which is called the **flow of execution**.

* Execution always begins at the **first statement** of the program. Statements are run one at a time, in order from top to bottom.
* Function definitions do not alter the flow of execution of the program, but remember that statements inside the function don’t run until the function is called.
* A function call is like a detour in the flow of execution. Instead of going to the next statement, the flow jumps to the body of the function, runs the statements there, and then comes back to pick up where it left off.
* That sounds simple enough, until you remember that one function can call another. While in the middle of one function, the program might have to run the statements in another function. Then, while running that new function, the program might have to run yet another function!
* Fortunately, Python is good at keeping track of where it is, so each time a function completes, the program picks up where it left off in the function that called it. When it gets to the end of the program, it terminates.

In summary, when you read a program, **you don’t always want to read from top to bottom**. Sometimes it makes more sense if you follow the ****flow of execution.


### 3.7  Parameters and arguments

Some of the functions we have seen require arguments. 

For example, when you call math.sin you pass a number as an argument. Some functions take more than one argument: math.pow takes two, the base and the exponent.

Inside the function, the arguments are assigned to variables called parameters. Here is a definition for a function that takes an argument:

def print_twice(bruce):
    print(bruce)
    print(bruce)
    
This function assigns the argument to a parameter named bruce. When the function is called, it prints the value of the parameter (whatever it is) twice.

This function works with any value that can be printed.

>>> print_twice('Spam')
Spam
Spam
>>> print_twice(42)
42
42
>>> print_twice(math.pi)
3.14159265359
3.14159265359

The same rules of composition that apply to built-in functions also apply to programmer-defined functions, so we can use any kind of expression as an argument for print_twice:

>>> print_twice('Spam '*4)
Spam Spam Spam Spam
Spam Spam Spam Spam
    
>>> print_twice(math.cos(math.pi))
-1.0
-1.0
The argument is evaluated before the function is called, so in the examples the expressions 'Spam '*4 and math.cos(math.pi) are only evaluated once.

You can also use a variable as an argument:

>>> michael = 'Eric, the half a bee.'
>>> print_twice(michael)
Eric, the half a bee.
Eric, the half a bee.
The name of the variable we pass as an argument (michael) has nothing to do with the name of the parameter (bruce). It doesn’t matter what the value was called back home (in the caller); here in print_twice, we call everybody bruce.

My example:
>>> James = "James wake up!"
>>> print_twice(James)
James wake up!
James wake up!
>>> 

### 3.8  Variables and parameters are local

When you create a variable inside a function, it is local, which means that it only exists inside the function. 

For example:

def cat_twice(part1, part2):
    cat = part1 + part2
    print_twice(cat)
This function takes two arguments, **concatenates** them, and **prints** the result twice. Here is an example that uses it:

>>> line1 = 'Bing tiddle '
>>> line2 = 'tiddle bang.'
>>> cat_twice(line1, line2)
Bing tiddle tiddle bang.
Bing tiddle tiddle bang.
When cat_twice terminates, the variable cat is destroyed. If we try to print it, we get an exception:

>>> print(cat)
NameError: name 'cat' is not defined
Parameters are also local. For example, outside print_twice, there is no such thing as bruce.


### 3.9  Stack diagrams  **non-local**

* To keep track of which variables can be used where, it is sometimes useful to draw a **stack diagram**. 
* Like state diagrams, **stack diagrams** show the value of each variable, but they also show the function each variable belongs to.
* Each function is represented by a **frame**. A **frame** is a box with the name of a function beside it and the parameters and variables of the function inside it. The stack diagram for the previous example is shown in [Figure 3.1](http://www.greenteapress.com/thinkpython2/html/thinkpython2004.html).
* The frames are arranged in a stack that indicates which function called which, and so on. In this example, print_twice was called by cat_twice, and cat_twice was called by __main__, which is a special name for the topmost frame. When you create a variable **outside** of any function, it belongs to __main__.


* **Each parameter refers to the same value as its corresponding argument. So, part1 has the same value as line1, part2 has the same value as line2, and bruce has the same value as cat.**

* If an error occurs during a function call, Python prints the name of the function, the name of the function that called it, and the name of the function that called that, all the way back to __main__.

For example, if you try to access cat from within print_twice, you get a NameError:

```
Traceback (innermost last):
  File "test.py", line 13, in __main__
    cat_twice(line1, line2)
  File "test.py", line 5, in cat_twice
    print_twice(cat)
  File "test.py", line 9, in print_twice
    print(cat)
NameError: name 'cat' is not defined
```

This list of functions is called a traceback. It tells you what program file the error occurred in, and what line, and what functions were executing at the time. It also shows the line of code that caused the error.

* **The order of the functions in the traceback is the same as the order of the frames in the stack diagram. The function that is currently running is at the bottom.**


### 3.10  Fruitful functions and void functions

* Some of the functions we have used, such as the math functions, return results; for lack of a better name, I call them **fruitful functions**. 
* Other functions, like print_twice, perform an action but don’t return a value. They are called **void functions**.

```
e.g. Fruitful functions
When you call a fruitful function, you almost always want to do something with the result; for example, you might assign it to a variable or use it as part of an expression:

x = math.cos(radians)
golden = (math.sqrt(5) + 1) / 2
When you call a function in interactive mode, Python displays the result:

>>> math.sqrt(5)
2.2360679774997898
But in a script, if you call a fruitful function all by itself, the return value is lost forever!

```
math.sqrt(5)
This script computes the square root of 5, but since it doesn’t store or display the result, it is not very useful.

```
e.g. Void functions

Void functions might display something on the screen or have some other effect, but they don’t have a return value. If you assign the result to a variable, you get a special value called None.

>>> result = print_twice('Bing')
Bing
Bing

>>> print(result)
None

The value None is not the same as the string 'None'. It is a special value that has its own type:

>>> type(None)
<class 'NoneType'>
```
The functions we have written so far are all void. We will start writing **fruitful functions** in a few chapters.


### 3.11  Why functions?

It may not be clear why it is worth the trouble to divide a program into functions. There are several reasons:

    * Creating a new function gives you an opportunity to name a group of statements, which makes your program easier to read and debug.
    * Functions can make a program smaller by eliminating repetitive code. Later, if you make a change, you only have to make it in one place.
    * Dividing a long program into functions allows you to debug the parts one at a time and then assemble them into a working whole.
    *Well-designed functions are often useful for many programs. Once you write and debug one, you can reuse it.


### 3.12  Debugging

* One of the most important skills you will acquire is debugging. Although it can be frustrating, debugging is one of the most intellectually rich, challenging, and interesting parts of programming.

* In some ways debugging is like detective work. You are confronted with clues and you have to infer the processes and events that led to the results you see.

* Debugging is also like an experimental science. Once you have an idea about what is going wrong, you modify your program and try again. If your hypothesis was correct, you can predict the result of the modification, and you take a step closer to a working program. If your hypothesis was wrong, you have to come up with a new one. As Sherlock Holmes pointed out, “When you have eliminated the impossible, whatever remains, however improbable, must be the truth.” (A. Conan Doyle, The Sign of Four)

* For some people, programming and debugging are the same thing. That is, programming is the process of gradually debugging a program until it does what you want. The idea is that you should start with a working program and make small modifications, debugging them as you go.

For example, Linux is an operating system that contains millions of lines of code, but it started out as a simple program Linus Torvalds used to explore the Intel 80386 chip. According to Larry Greenfield, “One of Linus’s earlier projects was a program that would switch between printing AAAA and BBBB. This later evolved to Linux.” (The Linux Users’ Guide Beta Version 1).

### 3.13  Glossary

**function:**
A named sequence of statements that performs some useful operation. Functions may or may not take arguments and may or may not produce a result.

**function definition:**
A statement that creates a new function, specifying its name, parameters, and the statements it contains.

**function object:**
A value created by a function definition. The name of the function is a variable that refers to a function object.
 ```
 >>> def print_lyrics():
...     print("I'm a lumberjack, and I'm okay.")
...     print("I sleep all night and I work all day.")
...
To end the function, **you have to enter an empty line**.

Defining a function creates a **function object**.

In this case, the followings are function objects."
...     print("I'm a lumberjack, and I'm okay.")
...     print("I sleep all night and I work all day.")
```

**header:**
The first line of a function definition.

**body:**
The sequence of statements inside a function definition.

**parameter:**
A name used inside a function to refer to the value passed as an argument.

**function call:**
A statement that runs a function. It consists of the function name followed by an argument list in parentheses.

**argument:**
A value provided to a function when the function is called. This value is assigned to the corresponding parameter in the function.

```
>>> type(42)
<class 'int'>
The name of the function is type. The expression in parentheses is called the **argument** of the function. The result, for this function, is the type of the argument.

It is common to say that a function “takes” an argument and “returns” a result. The result is also called the return value.
```

**local variable:**
A variable defined **inside** a function. A local variable can only be used inside its function.

**return value:**
The result of a function. If a function call is used as an expression, the return value is the value of the expression.

**fruitful function:**
A function that returns a value.

**void function:**
A function that always returns None.

**None:**
A special value returned by void functions.

**module:**
A file that contains a collection of related functions and other definitions.

**import statement:**
A statement that reads a module file and creates a module object.

**module object:**
A value created by an import statement that provides access to the values defined in a module.

**dot notation:**
The syntax for calling a function in another module by specifying the module name followed by a dot (period) and the function name.

**composition:**
Using an expression as part of a larger expression, or a statement as part of a larger statement.

**flow of execution:**
The order statements run in.

**stack diagram:**
A graphical representation of a stack of functions, their variables, and the values they refer to.

**frame:**
A box in a stack diagram that represents a function call. It contains the local variables and parameters of the function.

**traceback:**
A list of the functions that are executing, printed when an exception occurs.

In [ ]:
### 3.14  Exercises


#### Exercise 1

Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

>>> right_justify('monty')
                                                                 monty
Hint: Use string concatenation and repetition. Also, Python provides a built-in function called len that returns the length of a string, so the value of len('monty') is 5.


def right_justify(part1, part2):
	monty = part1 + part2
	print(monty)
	
	line1 = ' '*69
	line2 = 'monty'
	right_justify(part1, part2)

#### Exercise 2  

A **function object** is a value you can assign to a variable or pass as an argument. 

For example, do_twice is a function that takes a **function object** as an argument and calls it twice:

def do_twice(f):
    f()
    f()

NK:"f" is an argument, and f() f() are function objects in this case.(see definitons above). 

Here’s an example that uses do_twice to call a function named print_spam twice.

def print_spam():
    print('spam')

do_twice(print_spam)

1. Type this example into a script and test it.

```
>>> def do_twice(f):
...     f()
...     f()
... 
>>> def print_spam():
...     print('spam')
... 
>>> do_twice(print_spam)
spam
spam
>>> 
```

2. Modify do_twice so that it takes two arguments, a function object and a value, and calls the function twice, passing the value as an argument.

```
A **value** is one of the basic things a program works with, like a **letter or a number**. Some values we have seen so far are 2, 42.0, and 'Hello, World!
```
```
def do_twice(func, arg):
    """Runs a function twice.

    func: function object
    arg: argument passed to the function nk: means func(arg)
    """
    func(arg)
    func(arg)
```

3. Copy the definition of print_twice from earlier in this chapter to your script.

```
def print_twice(bruce):
    print(bruce)
    print(bruce)
    
This function assigns the argument to a parameter named bruce. When the function is called, it prints the value of the parameter (whatever it is) twice.

This function works with any value that can be printed.

>>> print_twice('Spam')
Spam
Spam
>>> print_twice(42)
42
42
```



4. Use the modified version of do_twice to call print_twice twice, passing 'spam' as an argument.

```

>>> def print_twice(arg):
...     print(arg)
...     print(arg)
... 
>>> print_twice('spam')
spam
spam
>>> 
```

5. Define a new function called do_four that takes a function object and a value and calls the function four times, passing the value as a parameter. There should be only two statements in the body of this function, not four.

```
def do_four(func, arg):
    """Runs a function four times.

    func: function object
    arg: argument passed to the function
    """
    do_twice(func, arg)
    do_twice(func, arg)
    """
    NK: do_twice is function which already defined in the question 4, and func, and arg are   
    the arguments passed to the function.
    """

do_twice(print_twice, 'spam')
print('')

do_four(print_twice, 'spam')
```

NAOKO's run
```
>>> def do_twice(func, arg):
...     func(arg)
...     func(arg)
... 
>>> def print_twice(arg):
...     print(arg)
...     print(arg)
... 
>>> def do_four(func, arg):
...     do_twice(func, arg)
...     do_twice(func, arg)
... 
>>> do_twice(print_twice, 'spam')
spam
spam
spam
spam
>>> print('')

>>> do_four(print_twice, 'spam')
spam
spam
spam
spam
spam
spam
spam
spam
>>> 
```

Solution: http://thinkpython2.com/code/do_four.py.

### Exercise 3  

Note: This exercise should be done using only the statements and other features we have learned so far.

Write a function that draws a grid like the following:
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +

Hint: to print more than one value on a line, you can print a comma-separated sequence of values:

print('+', '-')

By default, print advances to the next line, but you can override that behavior and put a space at the end, like this:

print('+', end=' ')
print('-')
The output of these statements is '+ -' on the same line. The output from the next print statement would begin on the next line.

Write a function that draws a similar grid with four rows and four columns.

Answer example:

```
from __future__ import print_function, division

# here is a mostly-straightforward solution to the two-by-two version of the grid.

def do_twice(f):
    f()
    f()

def do_four(f):
    do_twice(f)
    do_twice(f)

def print_beam():
    print('+ - - - -', end=' ')

def print_post():
    print('|        ', end=' ')

def print_beams():
    do_twice(print_beam)
    print('+')

def print_posts():
    do_twice(print_post)
    print('|')

def print_row():
    print_beams()
    do_four(print_posts)

def print_grid():
    do_twice(print_row)
    print_beams()

print_grid()
    
```

## Chapter 4  Case study: interface design

* This chapter presents a case study that demonstrates a **process for designing functions that work together**.

* It introduces the turtle module, which allows you to create **images** using turtle graphics. 
* The turtle module is included in most Python installations, but if you are running Python using PythonAnywhere, you won’t be able to run the turtle examples (at least you couldn’t when I wrote this).

If you have already installed Python on your computer, you should be able to run the examples. Otherwise, now is a good time to install. I have posted instructions at http://tinyurl.com/thinkpython2e.

Code examples from this chapter are available from http://thinkpython2.com/code/polygon.py.

### 4.1  The turtle module

To check whether you have the turtle module, open the Python interpreter and type

```
>>> import turtle
>>> bob = turtle.Turtle()

```
When you run this code, it should create a new window with small arrow that represents the turtle. Close the window.


Create a file named mypolygon.py 

```
Teachings-MacBook-Air:~ teachinghouse$ mkdir pythonpractice
Teachings-MacBook-Air:~ teachinghouse$ cd pythonpractice/
Teachings-MacBook-Air:pythonpractice teachinghouse$ touch mypolygon.py
Teachings-MacBook-Air:pythonpractice teachinghouse$ ls 
mypolygon.py

##The touch command is the easiest way to create new, empty files.

```

Type in the following code:

```
import turtle
bob = turtle.Turtle()
print(bob)
turtle.mainloop()
```

The turtle module (with a lowercase ’t’) provides a **function called Turtle** (with an uppercase ’T’) that creates a Turtle object, which we assign to a variable named bob. Printing bob displays something like:

<turtle.Turtle object at 0xb7bfbf4c>

This means that bob refers to an object with type Turtle as defined in module turtle.

mainloop tells the window to wait for the user to do something, although in this case there’s not much for the user to do except **close the window** (or control + Z).

Once you create a Turtle, you can call a method to move it around the window. A **method** is similar to a function, but it uses slightly different syntax. 

For example, to move the turtle forward:

```
bob.fd(100)
```

The method, fd, is associated with the turtle object we’re calling bob. **Calling a method is like making a request**: you are asking bob to move forward.

The argument of fd is a distance in **pixels**, so **the actual size depends on your display**.

Other methods you can call on a Turtle are:
* **bk to move backward**
* **lt for left turn**
* **rt right turn**. 

The argument for lt and rt is an angle in degrees.

Also, each Turtle is holding a pen, which is either down or up; if the pen is down, the Turtle leaves a trail when it moves. The methods **pu** and **pd** stand for **“pen up” and “pen down”**.

To draw a right angle, add these lines to the program (after creating bob and before calling mainloop):

```
import turtle
bob = turtle.Turtle()
print(bob)
bob.fd(100)
bob.lt(90)
bob.fd(100)
```
When you run this program, you should see bob move east and then north, leaving two line segments behind.

Now modify the program to draw a square. Don’t go on until you’ve got it working!

Additionally types:

```
bob.lt(90)
bob.fd(100)
bob.lt(90)
bob.fd(100)
```

### 4.2  Simple repetition

Chances are you wrote something like this:

bob.fd(100)
bob.lt(90)

bob.fd(100)
bob.lt(90)

bob.fd(100)
bob.lt(90)

bob.fd(100)

We can do the same thing more concisely with a **for** statement. Add this example to mypolygon.py and run it again:

for i in range(4):
    print('Hello!')
You should see something like this:

Hello!
Hello!
Hello!
Hello!

```
Teachings-MacBook-Air:pythonpractice teachinghouse$ ls
mypolygon.py
Teachings-MacBook-Air:pythonpractice teachinghouse$ echo "for i in range(4):" > mypolygon.py 
Teachings-MacBook-Air:pythonpractice teachinghouse$ echo "    print('Hello!')" >> mypolygon.py 
-bash: !': event not found
Teachings-MacBook-Air:pythonpractice teachinghouse$ echo "    print('Hello')" >> mypolygon.py 
Teachings-MacBook-Air:pythonpractice teachinghouse$ cat mypolygon.py 
for i in range(4):
    print('Hello')
Teachings-MacBook-Air:pythonpractice teachinghouse$ python mypolygon.py  
Hello
Hello
Hello
Hello

#
#run a program by invoking the Python interpreter by typing in bash:
python XXX.py
```


This is the simplest use of the for statement; we will see more later. But that should be enough to let you rewrite your square-drawing program. Don’t go on until you do.

Here is a for statement that draws a square:

```
for i in range(4):
    bob.fd(100)
    bob.lt(90)
```

    
* The syntax of a for statement is similar to a function definition. It has a header that ends with a colon and an indented body. The body can contain any number of statements.

* A **for** statement is also called a **loop** because the flow of execution runs through the body and then loops back to the top. In this case, it runs the body four times.

* This version is actually a little different from the previous square-drawing code because it makes another turn after drawing the last side of the square. The extra turn takes more time, but it simplifies the code if we do the same thing every time through the loop. This version also has the effect of leaving the turtle back in the starting position, facing in the starting direction.

### 4.3  Exercises

* The following is a series of exercises using TurtleWorld. They are meant to be fun, but they have a point, too. While you are working on them, think about **what the point is**.

* The following sections have solutions to the exercises, so don’t look until you have finished (or at least tried).

1. Write a function called square that takes a parameter named t, which is a turtle. It should use the turtle to draw a square.
Write a function call that passes bob as an argument to square, and then run the program again.

2. Add another parameter, named length, to square. Modify the body so length of the sides is length, and then modify the function call to provide a second argument. Run the program again. Test your program with a range of values for length.

3. Make a copy of square and change the name to polygon. Add another parameter named n and modify the body so it draws an n-sided regular polygon. Hint: The exterior angles of an n-sided regular polygon are 360/n degrees.

4. Write a function called circle that takes a turtle, t, and radius, r, as parameters and that draws an approximate circle by calling polygon with an appropriate length and number of sides. Test your function with a range of values of r.
Hint: figure out the circumference of the circle and make sure that length * n = circumference.

5. Make a more general version of circle called arc that takes an additional parameter angle, which determines what fraction of a circle to draw. angle is in units of degrees, so when angle=360, arc should draw a complete circle.


### 4.4  Encapsulation

The first exercise asks you to put your square-drawing code into a function definition and then call the function, passing the turtle as a parameter. Here is a solution:

```
def square(t):
    for i in range(4):
        t.fd(100)
        t.lt(90)
square(bob)

#"i" will do __t.fd(100)__and __t.lt(90)__ once. range(4) means repeat "i" 4 times. 
```

* The innermost statements, fd and lt are indented twice to show that they are inside the **for loop**, which is inside the function definition (function is "square", indented part is called the function defenition. 

* The next line, square(bob), is flush with the left margin, which indicates the end of both the for loop and the function definition.

* Inside the function, t refers to the same turtle bob, so t.lt(90) has the same effect as bob.lt(90). In that case, why not call the parameter bob? **The idea is that t can be any turtle, not just bob**, so you could create a second turtle named alice and pass it as an argument to square:

```
alice = turtle.Turtle()
square(alice)
```

Wrapping a piece of code up in a function is called **encapsulation**. 
Benefits of encapsulation:
1. encapsulation attaches a name to the code, which serves as a kind of documentation. 
2. Another advantage is that if you re-use the code, **it is more concise to call a function twice than to copy and paste the body!**


### 4.5  Generalization

The next step is to add a **length** parameter to square. Here is a solution:

```
def square(t, length):
    for i in range(4):
        t.fd(length)
        t.lt(90)

square(bob, 100)
```

* Adding a parameter to a function is called **generalization** because **it makes the function more general**: in the previous version, the square is always the same size; in this version it can be **any size**.

* The next step is also a generalization. Instead of drawing squares, polygon draws regular polygons with any number of sides. Here is a solution:

```
def polygon(t, n, length):
    angle = 360 / n
    for i in range(n):
        t.fd(length)
        t.lt(angle)

polygon(bob, 7, 70)
```
polygon(bob, 6, 70)

This example draws a 7-sided polygon with side length 70.

* If you are using Python 2, the value of angle might be off because of integer division. A simple solution is to compute angle = 360.0 / n. Because the numerator is a floating-point number, the result is floating point.

* When a function has more than a few numeric arguments, it is easy to forget what they are, or what order they should be in. In that case it is often a good idea to include the names of the parameters in the argument list:

```
polygon(bob, n=7, length=70)

```

These are called keyword arguments because they include the parameter names as “keywords” (not to be confused with Python keywords like while and def).

This syntax makes the program more readable. It is also a reminder about how arguments and parameters work: when you call a function, the arguments are assigned to the parameters.


### 4.6  Interface design

The next step is to write circle, which takes a radius, r, as a parameter. Here is a simple solution that uses polygon to draw a 50-sided polygon:

```
import math

def circle(t, r):
    circumference = 2 * math.pi * r
    n = 50
    length = circumference / n
    polygon(t, n, length)
``` 

* The first line computes the circumference of a circle with radius r using the formula 2 π r. Since we use math.pi, we have to import math. By convention, import statements are usually at the beginning of the script.

* n is the number of line segments in our approximation of a circle, so length is the length of each segment. Thus, polygon draws a 50-sided polygon that approximates a circle with radius r.

* One limitation of this solution is that n is a constant, which means that for very big circles, the line segments are too long, and for small circles, we waste time drawing very small segments. One solution would be to generalize the function by taking n as a parameter. This would give the user (whoever calls circle) more control, but the interface would be less clean.

* The interface of a function is a summary of how it is used: what are the parameters? What does the function do? And what is the return value? An interface is “clean” if it allows the caller to do what they want without dealing with unnecessary details.

* In this example, r belongs in the interface because it specifies the circle to be drawn. n is less appropriate because it pertains to the details of how the circle should be rendered.

* Rather than clutter up the interface, it is better to choose an appropriate value of n depending on circumference:

```
def circle(t, r):
    circumference = 2 * math.pi * r
    n = int(circumference / 3) + 3
    length = circumference / n
    polygon(t, n, length)
```
circle(bob, 100)

* Now the number of segments is an integer near circumference/3, so the length of each segment is approximately 3, which is small enough that the circles look good, but big enough to be efficient, and acceptable for any size circle.

Adding 3 to n guarantees that the polygon has at least 3 sides.

### 4.7  Refactoring

* **When I wrote circle, I was able to re-use polygon because a many-sided polygon is a good approximation of a circle**. But arc is **not as cooperative**; **we can’t use polygon or circle to draw an arc**.

* One alternative is to start with a copy of polygon and transform it into arc. The result might look like this:


```
def arc(t, r, angle):
    arc_length = 2 * math.pi * r * angle / 360
    n = int(arc_length / 3) + 1
    step_length = arc_length / n
    step_angle = angle / n
    
    for i in range(n):
        t.fd(step_length)
        t.lt(step_angle)
```
The second half of this function looks like polygon, but we can’t re-use polygon without changing the interface. We could generalize polygon to take an angle as a third argument, but then polygon would no longer be an appropriate name! Instead, let’s call the more general function polyline:

```
def polyline(t, n, length, angle):
    for i in range(n):
        t.fd(length)
        t.lt(angle)
```
Now we can rewrite polygon and arc to use polyline:

```
def polygon(t, n, length):
    angle = 360.0 / n
    polyline(t, n, length, angle)

def arc(t, r, angle):
    arc_length = 2 * math.pi * r * angle / 360
    n = int(arc_length / 3) + 1
    step_length = arc_length / n
    step_angle = float(angle) / n
    polyline(t, n, step_length, step_angle)
    
```

Finally, we can rewrite circle to use arc:

```
def circle(t, r):
    arc(t, r, 360)
    
```

* This process—rearranging a program to improve interfaces and facilitate code re-use—is called refactoring. In this case, we noticed that there was similar code in arc and polygon, so we “factored it out” into polyline.

* If we had planned ahead, we might have written polyline first and avoided refactoring, but often you don’t know enough at the beginning of a project to design all the interfaces. Once you start coding, you understand the problem better. Sometimes refactoring is a sign that you have learned something.

### 4.8  A development plan

* A development plan is a process for writing programs. The process we used in this case study is “encapsulation and generalization”. The steps of this process are:

1. Start by writing a small program with no function definitions.
2. Once you get the program working, identify a coherent piece of it, encapsulate the piece in a function and give it a name.
3. Generalize the function by adding appropriate parameters.
4. Repeat steps 1–3 until you have a set of working functions. Copy and paste working code to avoid retyping (and re-debugging).
5. Look for opportunities to improve the program by refactoring. For example, if you have similar code in several places, consider factoring it into an appropriately general function.


This process has some drawbacks—we will see alternatives later—but it can be useful if you don’t know ahead of time how to divide the program into functions. This approach lets you design as you go along.

### 4.9  docstring

A **docstring** is a string at the beginning of a function that ***explains the interface*** (“doc” is short for “documentation”). Here is an example:

```
def polyline(t, n, length, angle):
    """Draws n line segments with the given length and
    angle (in degrees) between them.  t is a turtle.
    """    
    for i in range(n):
        t.fd(length)
        t.lt(angle)
```
* By convention, all docstrings are triple-quoted strings, also known as multiline strings because the triple quotes allow the string to span more than one line.

* It is terse, but it contains the essential information someone would need to use this function. It explains concisely what the function does (without getting into the details of how it does it). It explains what effect each parameter has on the behavior of the function and what type each parameter should be (if it is not obvious).

* Writing this kind of documentation is an important part of interface design. A well-designed interface should be simple to explain; if you have a hard time explaining one of your functions, maybe the interface could be improved.

### 4.10  Debugging

* An interface is like a contract between a function and a caller. The caller agrees to provide certain parameters and the function agrees to do certain work.

For example, polyline requires four arguments: t has to be a Turtle; n has to be an integer; length should be a positive number; and angle has to be a number, which is understood to be in degrees.

* These requirements are called preconditions because they are supposed to be true before the function starts executing. Conversely, conditions at the end of the function are postconditions. Postconditions include the intended effect of the function (like drawing line segments) and any side effects (like moving the Turtle or making other changes).

* **Preconditions** are the responsibility of the caller. If the caller violates a (properly documented!) precondition and the function doesn’t work correctly, the bug is in the caller, not the function.

* If the preconditions are satisfied and the postconditions are not, the bug is in the function. If your pre- and postconditions are clear, they can help with debugging.

### 4.11  Glossary

**method:**
A function that is associated with an object and called using dot notation.

**loop:**
A part of a program that can run repeatedly.

**encapsulation:**
The process of transforming a sequence of statements into a function definition.

**generalization:**
The process of replacing something unnecessarily specific (like a number) with something appropriately general (like a variable or parameter).

**keyword argument:**
An argument that includes the name of the parameter as a “keyword”.

**interface:**
A description of how to use a function, including the name and descriptions of the arguments and return value.

**refactoring:**
The process of modifying a working program to improve function interfaces and other qualities of the code.

**development plan:**
A process for writing programs.

**docstring:**
A string that appears at the top of a function definition to document the function’s interface.

**precondition:**
A requirement that should be satisfied by the caller before a function starts.

**postcondition:**
A requirement that should be satisfied by the function before it ends.
